classes/PaymentModule.php

```diff
@@ -1267,4 +1272,19 @@ protected function createOrderCartRules(
 
         return $cart_rules_list;
     }
+
+    protected function isCarrierValid($package_list): bool
+    {
+        foreach ($package_list as $packageByAddress) {
+            foreach ($packageByAddress as $package) {
+                $carrier = new Carrier((int) $package['id_carrier']);
```

Comment:
@pszostok it seems that this part fails when you create an order from back office, could you check information from the tests and update your code accordingly?

PS: dzięki za reakcje na tego buga :)

Reply:
@kpodemski Thank you for pointing out the errors in the tests. I am aware of them, but unfortunately, I haven't had the time to address them yet. As soon as I can, I will analyze the tests and make the necessary fixes in my PR to resolve the issue.

Reply:
@pszostok During my tests, the `id_carrier` property was not always present. I am not sure whether this is dependent on PrestaShop version or some other factor, but current state of the fix would never allow any order to complete in such circumstance. I suggest adding more variable checks and some other method to detect the carrier if `id_carrier` is not set.

My proposal below. I'd commit this myself, but I don't have write access to your fork. (Invitation expired. 🥲)
I'd also suggest considering removing `id_carrier` path entirely, if `$card_delivery_option` path is more robust. But PrestaShop adding `,` at the end of the carrier identifiers does not bode well.

```php
if (!$this->isCarrierValid($package_list, $cart_delivery_option)) {
    PrestaShopLogger::addLog('PaymentModule::validateOrder - Carrier is not valid', 3, null, 'Cart', (int) $id_cart, true);
    die(Tools::displayError('Error processing order. Carrier is not valid.'));
}

//...

protected function isCarrierValid($package_list, $cart_delivery_option): bool
{
    foreach ($package_list as $packageKey => $packageByAddress) {
        foreach ($packageByAddress as $package) {
            if (isset($package['id_carrier'])) {
                $carrier = new Carrier((int) $package['id_carrier']);
            } else {
                $carrier = new Carrier((int) reset(explode(',', $cart_delivery_option[$packageKey])));
            }

            if (!empty($carrier)) {
                if (empty(Db::getInstance()->executeS('SELECT `id_module` FROM `' . _DB_PREFIX_ . 'module_carrier` WHERE `id_reference` = ' . (int) $carrier->id_reference . ' AND `id_shop` = ' . (int) $this->context->shop->id . ' AND `id_module` = ' . (int) $this->id))) {
                    return false;
                }
            }
        }
    }

    return true;
}
```

Reply:
The comma is there for a reason, the deliver option is a COLLECTION of the carriers inside that delivery option. There can be multiple carriers for one delivery option - multishipping.

If you want to do something with carriers, it's best to use carrier_list inside the delivery option.

Reply:
Ok, that makes sense! Thank you for the added context.

I would imagine a list would be an array, but I guess every codebase has its legacy issues.

We'll reevaluate accordingly and then update the PR. 🙂